gio/src/cancellable_future.rs

```diff
@@ -87,49 +149,37 @@ where
     type Output = Result<<F as Future>::Output, Cancelled>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        if self.is_cancelled() {
-            return Poll::Ready(Err(Cancelled));
+        if self.cancellable.is_cancelled() {
+            // XXX: Whenever we want to break the API, we should return here only
+            // self.cancellable.set_error_if_cancelled() value.
+            return std::task::Poll::Ready(Err(Cancelled));
         }
 
-        let mut this = self.as_mut().project();
+        let mut waker = self.waker.lock().unwrap();
+        if waker.is_none() {
+            *waker = Some(cx.waker().clone());
+        }
+        drop(waker);
 
+        let this = self.as_mut().project();
         match this.future.poll(cx) {
             Poll::Ready(out) => Poll::Ready(Ok(out)),
-
-            Poll::Pending => {
-                if let Some(prev_handler) = this.waker_handler_cb.take() {
-                    this.cancellable.disconnect_cancelled(prev_handler);
-                }
-
-                let canceller_handler_id = this.cancellable.connect_cancelled({
-                    let w = cx.waker().clone();
-                    move |_| w.wake()
-                });
-
-                match canceller_handler_id {
-                    Some(canceller_handler_id) => {
-                        *this.waker_handler_cb = Some(canceller_handler_id);
-                        Poll::Pending
-                    }
-
-                    None => Poll::Ready(Err(Cancelled)),
-                }
-            }
+            Poll::Pending => Poll::Pending,
         }
     }
 }
 
 impl From<Cancelled> for glib::Error {
     fn from(_: Cancelled) -> Self {
-        glib::Error::new(IOErrorEnum::Cancelled, "Task cancelled")
+        glib::Error::new(IOErrorEnum::Cancelled, "Operation was cancelled")
     }
 }
 
 impl std::error::Error for Cancelled {}
```

Comment:
I would move the `XXX:...` comment of yours here, mentioning that maybe we can get rid of a specific error and just wrap the "cancel" error into a glib::Error. With API break of coure